README.md
inGenerator Source cookbook
=================================
[![Build Status](https://travis-ci.org/ingenerator/chef-ingenerator-source.png?branch=1.x)](https://travis-ci.org/ingenerator/chef-ingenerator-source)

The `ingenerator-source` cookbook provides a standard framework to manage deployment and provisioning of project source
code on an instance. This is used either for atomic checkouts on a live instance, or for running the same provisioning
and build code for a checkout on a development VM or build server.

Requirements
------------
- Chef 12.13 or higher
- **Ruby 2.3 or higher**

Installation
------------
We recommend adding to your `Berksfile` and using [Berkshelf](http://berkshelf.com/):

```ruby
cookbook 'ingenerator-source', git: 'git://github.com/ingenerator/chef-ingenerator-source', branch: 'master'
```

Have your main project cookbook *depend* on ingenerator-source by editing the `metadata.rb` for your cookbook.

```ruby
# metadata.rb
depends 'ingenerator-source'
```

Usage
-----
There's probably a fair bit of 'deployment' you still need to do on a local VM to get it up and running. You may need to
build assets, provision config files with details of database and API services, install composer dependencies etc. Much
of this provisioning is likely to overlap with the deployment/provisioning workflow on a live or build server. So this
cookbook wraps both types of operations together to make it easy to share details between both.

Your runlist just needs to include the `ingenerator-source::default` recipe and the rest of the process is controlled by
attributes.

### `:in_place` provisioning - eg on a VM

With in-place provisioning, the recipe simply creates a symlink from your deploy destination to the source path you
specify. There are no revision checkout folders, and if the source path is a vagrant share or similar you'll have a live
checkout. Once the symlink is in place the post-deploy hooks will execute as though following a deploy.

| attribute                                | default                         | notes                                   |
|------------------------------------------|---------------------------------|-----------------------------------------|
| node['project']['deploy']['type']        |                                 | set to `:in_place` to trigger this mode |
| node['project']['deploy']['source']      | /vagrant                        | the location of your raw source files   |
| node['project']['deploy']['destination'] | /var/www/{project_name}/current | the base web path for your application  |

### `:deploy` - eg on a build or live server

This is the default action.

The standard workflow is to check out the revision you want in a source repository on the instance - to get up to date
versions of the provisioning scripts etc - and then allow this cookbook to deploy a known checkout from there to the
web root using the chef deploy resource. Again, post-deploy hooks will execute at key points in the process.

| attribute                                | default                         | notes                                     |
|------------------------------------------|---------------------------------|-------------------------------------------|
| node['project']['deploy']['type']        | :deploy                         | set to `:deploy` to trigger this mode     |
| node['project']['deploy']['force']       | false                           | whether to force a redeploy of same rev   |
| node['project']['deploy']['source']      | /var/www/www-source             | the location of your raw source files     |
| node['project']['deploy']['destination'] | /var/www/{project_name}/current | the base web path for your application    |
| node['project']['deploy']['revision']    | HEAD                            | you should checkout the desired rev first |

### Marking known-unstable branches

You can prevent branches that you know to be unstable from making it into production accidentally. For example, if you are
part way through heavy refactoring of a legacy system and know there are problems that are not yet caught be your
unit/acceptance tests.

If the node's `project.deploy.allow_blocked_branch` is false (default) then the deploy run will throw an exception if a
PROD_RELEASE_BLOCKER file is found in the checkout root directory. You should add relevant information about why the
branch is blocked into this file.

If the `project.deploy.allow_blocked_branch` is true then the deploy run will issue a warning but continue. Build slaves
and potentially QA should have this attribute set to true.

### Deployment hooks

You can provide recipes that run as callbacks before the checked out code is symlinked into
production, or immediately after the deploy completes. These have full access your recipes,
resources and the rest of the chef environment but note are *not* loaded until during the deployment
so they are not compiled as part of the main chef compile and in particular cannot include changes
to attributes, runlists etc that need to apply outside their own scope.

> *Caution: Use the correct checkout directory in your hooks*
> During a live deployment, the on_prepare callback is triggered before the symlink to your code has been updated.
> You can get the path to the about-to-be-live code with the `this_release_path` helper method.
> This will throw if you attempt to access it before the prepare_deploy stage of the deployment.
>
> The on_complete hook can target either `this_release_path` or `node['project']['deploy']['destination']`
> as these paths resolve the same by this point.

| attribute                       | default | called                                             |
|---------------------------------|---------|----------------------------------------------------|
| `project.deploy.on_prepare`     | nil     | after all files are checked out, before symlinking |
| `project.deploy.on_complete`    | nil     | after deploy linked into production                |


```ruby
 # Specify your prepare recipe
 default['project']['deploy']['on_prepare'] = 'project::prepare_deploy'

 # project/recipes/prepare_deploy.rb
 composer_dependencies this_release_path do
   run_as  node['apache']['user']
 end
```

You can obviously make steps within your callbacks conditional on node attributes, etc, if required.

### Testing
See the [.travis.yml](.travis.yml) file for the current test scripts.

Contributing
------------
1. Fork the project
2. Create a feature branch corresponding to your change
3. Create specs for your change
4. Create your changes
4. Create a Pull Request on github

License & Authors
-----------------
- Author:: Andrew Coulton (andrew@ingenerator.com)

```text
Copyright 2012-2013, inGenerator Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/). Note that
0.x versions may be breaking, per the semver standard.

## Unreleased

## 2.0.0 (2017-08-10)

* Exclude unnecessary build-time files and dependencies from the vendored cookbook
* Update build dependencies and build against Chef 12 and Chef 13

## 1.0.0 (2017-02-24)

* [BREAKING] Remove the default prepare_deploy and complete_deploy recipes - these
  were only being used to verify the path was set and the new helper does that
  better.
* [BREAKING] Current release path is no longer set in node.deploy.release_path -
  instead you must use `this_release_path` which will throw if it hasn't been
  set (because deployment hasn't happened yet) when called.
* [BREAKING] No longer attempts to automatically install composer packages for
  a prepare_deploy - do this in your project if you need it.

## 0.1.0 (2017-02-24)

* First tagged release


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'
require 'support/matchers.rb'

RSpec.configure do |c|
  c.filter_run(focus: true)
  c.run_all_when_everything_filtered = true

  # Default platform / version to mock Ohai data from
  c.platform = 'ubuntu'
  c.version = '14.04'

  # Don't clear cookbooks from the server-runner between each test for performance
  c.server_runner_clear_cookbooks = false

end


spec/default_spec.rb
require 'spec_helper'

describe 'ingenerator-source::default' do
  let (:chef_run) { ChefSpec::SoloRunner.new.converge described_recipe }

  context 'by default' do
    it "uses the full deployment recipe" do
      expect(chef_run.node['project']['deploy']['type']).to be(:deploy)
    end

    it "sets the deploy destination to /var/www/{project_name}" do
      chef_run = ChefSpec::SoloRunner.new do |node|
        node.normal['project']['name'] = 'thisproject'
      end.converge(described_recipe)

      expect(chef_run.node['project']['deploy']['destination']).to eq('/var/www/thisproject')
    end

    it "sets the checkout owner to www-data" do
      expect(chef_run.node['project']['deploy']['owner']).to eq('www-data')
    end

    it "sets the checkout group to www-data" do
      expect(chef_run.node['project']['deploy']['group']).to eq('www-data')
    end

    it "attaches ingenerator-source::prepare_deploy as the on_prepare hook" do
      expect(chef_run.node['project']['deploy']['on_prepare']).to eq('ingenerator-source::prepare_deploy')
    end

    it "attaches ingenerator-source::complete_deploy as the on_prepare hook" do
      expect(chef_run.node['project']['deploy']['on_complete']).to eq('ingenerator-source::complete_deploy')
    end

  end

  context 'when node[project][deploy] is :in_place' do
    let (:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.normal['project']['deploy']['type'] = :in_place
      end.converge(described_recipe)
    end

    it 'should run the deploy_in_place recipe' do
      expect(chef_run).to include_recipe 'ingenerator-source::deploy_in_place'
    end
  end

  context 'when node[project][deploy] is :deploy' do
    let (:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.normal['project']['deploy']['type'] = :deploy
      end.converge(described_recipe)
    end

    it 'should run the deploy_revision recipe' do
      expect(chef_run).to include_recipe 'ingenerator-source::deploy_revision'
    end
  end

  context 'when node[project][deploy] is \'in_place\' from json' do
    let (:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.normal['project']['deploy']['type'] = 'in_place'
      end.converge(described_recipe)
    end

    it 'should run the deploy_in_place recipe' do
      expect(chef_run).to include_recipe 'ingenerator-source::deploy_in_place'
    end
  end

  context 'when node[project][deploy] is \'deploy\' from json' do
    let (:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.normal['project']['deploy']['type'] = 'deploy'
      end.converge(described_recipe)
    end

    it 'should run the deploy_revision recipe' do
      expect(chef_run).to include_recipe 'ingenerator-source::deploy_revision'
    end
  end

  context 'when node[project][deploy] is invalid' do
    let (:chef_run) do
      ChefSpec::SoloRunner.new do |node|
        node.normal['project']['deploy']['type'] = 'something random'
      end.converge(described_recipe)
    end

    it 'should throw an exception' do
      expect { chef_run }.to raise_error RuntimeError, /Invalid deploy type/
    end
  end

end


spec/support/matchers.rb
# Custom matchers for resources that don't define their own

def install_composer_project(project_path)
  ChefSpec::Matchers::ResourceMatcher.new(:composer_project, :install, project_path)
end


recipes/default.rb
#
# Deploys application source code, either locally or for a proper deployment
#
# Author::  Andrew Coulton (<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Recipe:: default
#
# Copyright 2012-13, inGenerator Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if (node['project'] && node['project']['name'])
  node.default['project']['deploy']['destination'] = "/var/www/#{node['project']['name']}"
end

if node['project']['deploy']['type'].to_s == 'in_place'
  include_recipe "ingenerator-source::deploy_in_place"

elsif node['project']['deploy']['type'].to_s == 'deploy'
  include_recipe "ingenerator-source::deploy_revision"

else
  raise "Invalid deploy type #{node['project']['deploy']['type']}"
end


attributes/defaults.rb
#
# Author:: Andrew Coulton(<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Attribute:: defaults
#
# Copyright 2012-13, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#


default['project']['deploy']['type']                 = :deploy
default['project']['deploy']['force']                = false
# This will be reset by the default recipe if a project.name attribute is present
default['project']['deploy']['destination']          = '/var/www/ingenerator-project'
# This will be set by the appropriate deploy recipe
default['project']['deploy']['source']               = nil

default['project']['deploy']['owner']                = 'www-data'
default['project']['deploy']['group']                = 'www-data'

default['project']['deploy']['allow_blocked_branch'] = false
default['project']['deploy']['on_prepare']           = nil
default['project']['deploy']['on_complete']          = nil


recipes/prepare_deploy.rb
#
# Default prepare_deploy recipe, called once the application has been checked out but before
# it is symlinked into production.
#
# Author::  Andrew Coulton (<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Recipe:: prepare_deploy
#
# Copyright 2012-13, inGenerator Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

unless node['project']['deploy']['release_path']
  raise(
    "The prepare_deploy recipe requires the node.project.deploy.release_path attribute to be set.\n"+
    "Are you trying to include the recipe directly rather than as a deploy callback?"
  )
end


recipes/complete_deploy.rb
#
# Default complete_deploy recipe, called after the application has been symlinked into production and
# is otherwise configured.
#
# Author::  Andrew Coulton (<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Recipe:: complete_deploy
#
# Copyright 2012-13, inGenerator Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

unless node['project']['deploy']['release_path']
  raise(
    "The prepare_deploy recipe requires the node.project.deploy.release_path attribute to be set.\n"+
    "Are you trying to include the recipe directly rather than as a deploy callback?"
  )
end


recipes/deploy_in_place.rb
#
# Creates a symlink from the web root to the source location and runs the post-deploy callbacks
#
# Author::  Andrew Coulton (<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Recipe:: default
#
# Copyright 2012-13, inGenerator Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Default deploy source for this kind of deployment
node.default['project']['deploy']['source'] = '/vagrant'

directory node['project']['deploy']['destination'] do
  recursive true
  user      node['project']['deploy']['owner']
  group     node['project']['deploy']['group']
  mode      0755
end

Ingenerator::SourceDeployment::DeploymentManager.set_release_path(node['project']['deploy']['destination']+'/current')
link this_release_path do
  to    node['project']['deploy']['source']
  user  node['project']['deploy']['owner']
  group node['project']['deploy']['group']
end


# Run the deploy hooks
%w(on_prepare on_complete).each do | hook |
  include_recipe node['project']['deploy'][hook] if node['project']['deploy'][hook]
end


libraries/deploy_helpers.rb
# Provides helper methods for interacting with the deployment during deploy
module Ingenerator
  module SourceDeployment
    class DeploymentManager
      # Assigned when the git checkout / etc is complete before symlinking
      @release_path = nil

      # Don't call this from external code
      def self.set_release_path(path)
        @release_path = path
      end

      class << self
        attr_reader :release_path
      end
    end

    module DeployHelpers
      class DeploymentNotReadyError < RuntimeError
        def initilize
          super
          "Your code attempted to access `this_release_path` before it had been set\n"\
          "This probably means you need to move your recipe later in execution\n"\
          '`this_release_path` is not set until the code has been checked out'
        end
      end

      # This method magically appears as a helper inside recipes, so you can
      # use it in your prepare_deploy hook recipe etc to access your checked-out
      # code before it goes live. If you call it before the code is there, it will
      # throw!
      def this_release_path
        path = Ingenerator::SourceDeployment::DeploymentManager.release_path
        raise DeploymentNotReadyError if path.nil?
        path
      end
    end
  end
end

Chef::Recipe.send(:include, Ingenerator::SourceDeployment::DeployHelpers)


recipes/deploy_revision.rb
#
# Runs the chef deployment resource and triggers any post-deploy handlers
#
# Author::  Andrew Coulton (<andrew@ingenerator.com>)
# Cookbook Name:: ingenerator-source
# Recipe:: default
#
# Copyright 2012-13, inGenerator Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Default deploy source for this kind of deployment
node.default['project']['deploy']['source'] = '/var/www/www-source'

# Do not allow known unstable branches into production
if File.exists?(node['project']['deploy']['source']+'/PROD_RELEASE_BLOCKER')
  if node['project']['deploy']['allow_blocked_branch']
    Chef::Log.warn('You are deploying a blocked branch - check the PROD_RELEASE_BLOCKER')
  else
    raise "Cannot deploy to production with a PROD_RELEASE_BLOCKER file - investigate and remove to continue"
  end
end

# Ensure deploy working directories exist
%w(releases shared).each do | path |
  directory "#{node['project']['deploy']['destination']}/#{path}" do
    recursive true
    owner     node['project']['deploy']['owner']
    group     node['project']['deploy']['group']
    mode      0755
  end
end

# Prevent hardlink problems with changing upstream references
directory "#{node['project']['deploy']['destination']}/shared/cached-copy" do
  action    :delete
  recursive true
end

deploy(node['project']['deploy']['destination']) do
  repo     node['project']['deploy']['source']
  action   (node['project']['deploy']['force'] ? :force_deploy : :deploy)
  revision node['project']['deploy']['revision']
  provider Chef::Provider::Deploy::Revision

  user     node['project']['deploy']['owner']
  group    node['project']['deploy']['group']

  # Clear this default railsy configuration, manage it in hooks
  symlink_before_migrate({})
  symlinks({})
  create_dirs_before_symlink []
  purge_before_symlink []

  # Run the hooks
  before_symlink do
    Ingenerator::SourceDeployment::DeploymentManager.set_release_path(release_path)
    if node['project']['deploy']['on_prepare']
      run_context.send :include_recipe, node['project']['deploy']['on_prepare']
    end
  end

  after_restart do
    if node['project']['deploy']['on_complete']
      run_context.send :include_recipe, node['project']['deploy']['on_complete']
    end
  end

end
